ecr-login/api/client.go

```diff
@@ -20,43 +20,61 @@ import (
 	"strings"
 	"time"
 
-	"github.com/pkg/errors"
-
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/ecr"
-	"github.com/awslabs/amazon-ecr-credential-helper/ecr-login/cache"
+	"github.com/aws/aws-sdk-go/service/ecrpublic"
+	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
+
+	"github.com/awslabs/amazon-ecr-credential-helper/ecr-login/cache"
 )
 
-const proxyEndpointScheme = "https://"
-const programName = "docker-credential-ecr-login"
+const (
+	proxyEndpointScheme = "https://"
+	programName         = "docker-credential-ecr-login"
+	ecrPublicName       = "public.ecr.aws"
+	ecrPublicEndpoint   = proxyEndpointScheme + ecrPublicName
+)
 
 var ecrPattern = regexp.MustCompile(`(^[a-zA-Z0-9][a-zA-Z0-9-_]*)\.dkr\.ecr(\-fips)?\.([a-zA-Z0-9][a-zA-Z0-9-_]*)\.amazonaws\.com(\.cn)?`)
 
+type Service string
+
+const (
+	ServiceECR       Service = "ecr"
+	ServiceECRPublic Service = "ecr-public"
+)
+
 // Registry in ECR
 type Registry struct {
-	ID     string
-	FIPS   bool
-	Region string
+	Service Service
+	ID      string
+	FIPS    bool
+	Region  string
 }
 
 // ExtractRegistry returns the ECR registry behind a given service endpoint
 func ExtractRegistry(serverURL string) (*Registry, error) {
 	if strings.HasPrefix(serverURL, proxyEndpointScheme) {
 		serverURL = strings.TrimPrefix(serverURL, proxyEndpointScheme)
 	}
+	if strings.HasPrefix(serverURL, ecrPublicName) {
```

Comment:
This check isn't sufficient as something like `public.ecr.aws.justkidding.example.com` would match.

Reply:
ecrPattern is having the same problem as well. I think it would be better to parse `serverURL` as a URL first and check specific components (scheme, host, ..) instead.

Reply:
Fixed, @kzys can you take a look?